pkg/sql/row/deleter.go

```diff
@@ -155,11 +163,87 @@ func (rd *Deleter) DeleteRow(
 		}
 		familyID := family.ID
 		rd.key = keys.MakeFamilyKey(primaryIndexKey, uint32(familyID))
-		if traceKV {
-			log.VEventf(ctx, 2, "Del %s", keys.PrettyPrint(rd.Helper.primIndexValDirs, rd.key))
+		if pkDeleter != nil {
+			prevValue, familyRequiresEncoding, err := rd.encodeValueForPrimaryIndexFamily(family, values)
+			if err != nil {
+				return err
+			}
+			if !familyRequiresEncoding {
+				return nil
+			}
+			var expValue []byte
+			if prevValue.IsPresent() {
+				expValue = prevValue.TagAndDataBytes()
+			}
+			pkDeleter(ctx, b, rd.key, expValue, traceKV)
+		} else {
+			if traceKV {
+				log.VEventf(ctx, 2, "Del %s", keys.PrettyPrint(rd.Helper.primIndexValDirs, rd.key))
+			}
+			b.Del(&rd.key)
 		}
-		b.Del(&rd.key)
+
 		rd.key = nil
 		return nil
 	})
 }
+
+// encodeValueForPrimaryIndexFamily encodes the expected roachpb.Value
+// for the given family and valuses.
+//
+// TODO(ssd): Lots of duplication between this and
+// prepareInsertAndUpdateBatch. This is rather unfortunate.
+func (rd *Deleter) encodeValueForPrimaryIndexFamily(
+	family *descpb.ColumnFamilyDescriptor, values []tree.Datum,
+) (roachpb.Value, bool, error) {
+	if len(family.ColumnIDs) == 1 && family.ColumnIDs[0] == family.DefaultColumnID && family.ID != 0 {
+		idx, ok := rd.FetchColIDtoRowIndex.Get(family.DefaultColumnID)
+		if !ok {
+			return roachpb.Value{}, false, nil
+		}
+		if rd.Helper.SkipColumnNotInPrimaryIndexValue(family.DefaultColumnID, values[idx]) {
+			return roachpb.Value{}, false, nil
+		}
+		typ := rd.FetchCols[idx].GetType()
+		marshaled, err := valueside.MarshalLegacy(typ, values[idx])
+		if err != nil {
+			return roachpb.Value{}, false, err
+		}
+
+		return marshaled, true, err
+	}
+
+	rd.rawValueBuf = rd.rawValueBuf[:0]
+	var lastColID descpb.ColumnID
+	familySortedColumnIDs, ok := rd.Helper.SortedColumnFamily(family.ID)
+	if !ok {
+		return roachpb.Value{}, false, errors.AssertionFailedf("invalid family sorted column id map")
+	}
+	for _, colID := range familySortedColumnIDs {
+		idx, ok := rd.FetchColIDtoRowIndex.Get(colID)
+		if !ok || values[idx] == tree.DNull {
```

Comment:
i think we need a close review of someone from queries around this logic. I think we'll need them to follow up and clean this up and unit test this.

Reply:
This is actually something like the 4th copy of pretty similar code that knows how to encode a set of values for a row. There is a version in `prepareInsertOrUpdateBatch`, `(*colenc.BatchEncoder).encodePK`, and then also `rowenc.EncodePrimaryIndex`.

Reply:
Hence:

> This is rather unfortunate.

I agree. Let's not loose track of cleaning this us. I filed an issue: https://github.com/cockroachdb/cockroach/issues/131589